### ISSUE TITLE
[BUG] GC stuck in loop trying to delete bucket after runing multipart integration test

### ISSUE THREAD

**Issue by 0xavi0** [s3gw-tech/s3gw#719]
**Describe the bug** 
The WAL checkpoint is being triggered constantly after running an integration test.
for example:
```bash
2023-09-20T16:34:47.950+0200 7ff344a636c0 10 [SQLITE] WAL checkpoint (passive) returned 0 (not an error), total_frames=1004, checkpointed_frames=1004
2023-09-20T16:34:48.080+0200 7ff344a636c0 10 [SQLITE] WAL checkpoint (passive) returned 0 (not an error), total_frames=1004, checkpointed_frames=1004
2023-09-20T16:34:48.217+0200 7ff344a636c0 10 [SQLITE] WAL checkpoint (passive) returned 0 (not an error), total_frames=1004, checkpointed_frames=1004
2023-09-20T16:34:48.350+0200 7ff344a636c0 10 [SQLITE] WAL checkpoint (passive) returned 0 (not an error), total_frames=1004, checkpointed_frames=1004
2023-09-20T16:34:48.494+0200 7ff344a636c0 10 [SQLITE] WAL checkpoint (passive) returned 0 (not an error), total_frames=1004, checkpointed_frames=1004
2023-09-20T16:34:48.624+0200 7ff344a636c0 10 [SQLITE] WAL checkpoint (passive) returned 0 (not an error), total_frames=1004, checkpointed_frames=1004
2023-09-20T16:34:48.754+0200 7ff344a636c0 10 [SQLITE] WAL checkpoint (passive) returned 0 (not an error), total_frames=1004, checkpointed_frames=1004
2023-09-20T16:34:48.887+0200 7ff344a636c0 10 [SQLITE] WAL checkpoint (passive) returned 0 (not an error), total_frames=1004, checkpointed_frames=1004
2023-09-20T16:34:49.020+0200 7ff344a636c0 10 [SQLITE] WAL checkpoint (passive) returned 0 (not an error), total_frames=1004, checkpointed_frames=1004
2023-09-20T16:34:49.154+0200 7ff344a636c0 10 [SQLITE] WAL checkpoint (passive) returned 0 (not an error), total_frames=1004, checkpointed_frames=1004
2023-09-20T16:34:49.290+0200 7ff344a636c0 10 [SQLITE] WAL checkpoint (passive) returned 0 (not an error), total_frames=1004, checkpointed_frames=1004
2023-09-20T16:34:49.420+0200 7ff344a636c0 10 [SQLITE] WAL checkpoint (passive) returned 0 (not an error), total_frames=1004, checkpointed_frames=1004
2023-09-20T16:34:49.560+0200 7ff344a636c0 10 [SQLITE] WAL checkpoint (passive) returned 0 (not an error), total_frames=1004, checkpointed_frames=1004
2023-09-20T16:34:49.700+0200 7ff344a636c0 10 [SQLITE] WAL checkpoint (passive) returned 0 (not an error), total_frames=1004, checkpointed_frames=1004
2023-09-20T16:34:49.834+0200 7ff344a636c0 10 [SQLITE] WAL checkpoint (passive) returned 0 (not an error), total_frames=1004, checkpointed_frames=1004
2023-09-20T16:34:49.964+0200 7ff344a636c0 10 [SQLITE] WAL checkpoint (passive) returned 0 (not an error), total_frames=1004, checkpointed_frames=1004
2023-09-20T16:34:50.104+0200 7ff344a636c0 10 [SQLITE] WAL checkpoint (passive) returned 0 (not an error), total_frames=1004, checkpointed_frames=1004
2023-09-20T16:34:50.240+0200 7ff344a636c0 10 [SQLITE] WAL checkpoint (passive) returned 0 (not an error), total_frames=1004, checkpointed_frames=1004
2023-09-20T16:34:50.377+0200 7ff344a636c0 10 [SQLITE] WAL checkpoint (passive) returned 0 (not an error), total_frames=1004, checkpointed_frames=1004
2023-09-20T16:34:50.517+0200 7ff344a636c0 10 [SQLITE] WAL checkpoint (passive) returned 0 (not an error), total_frames=1004, checkpointed_frames=1004
2023-09-20T16:34:50.664+0200 7ff344a636c0 10 [SQLITE] WAL checkpoint (passive) returned 0 (not an error), total_frames=1004, checkpointed_frames=1004
2023-09-20T16:34:50.794+0200 7ff344a636c0 10 [SQLITE] WAL checkpoint (passive) returned 0 (not an error), total_frames=1004, checkpointed_frames=1004
2023-09-20T16:34:50.934+0200 7ff344a636c0 10 [SQLITE] WAL checkpoint (passive) returned 0 (not an error), total_frames=1004, checkpointed_frames=1004
2023-09-20T16:34:51.067+0200 7ff344a636c0 10 [SQLITE] WAL checkpoint (passive) returned 0 (not an error), total_frames=1004, checkpointed_frames=1004
2023-09-20T16:34:51.204+0200 7ff344a636c0 10 [SQLITE] WAL checkpoint (passive) returned 0 (not an error), total_frames=1004, checkpointed_frames=1004
2023-09-20T16:34:51.344+0200 7ff344a636c0 10 [SQLITE] WAL checkpoint (passive) returned 0 (not an error), total_frames=1004, checkpointed_frames=1004
2023-09-20T16:34:51.481+0200 7ff344a636c0 10 [SQLITE] WAL checkpoint (passive) returned 0 (not an error), total_frames=1004, checkpointed_frames=1004
2023-09-20T16:34:51.611+0200 7ff344a636c0 10 [SQLITE] WAL checkpoint (passive) returned 0 (not an error), total_frames=1004, checkpointed_frames=1004
```
**To Reproduce** Steps to reproduce the behavior:

* start the radosgw
* run the multipart integration tests 
```bash
$ cd qa/rgw/store/sfs/tests
$ python -m unittest test-sfs-multipart.py
```
* stop the radosgw
* start the radosgw again

**Additional info**
After finishing the integration test the store folder looks like this:
```bash
$ ll /home/xavi/sfs
total 1.4M
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:43 34
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:43 3b
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:43 c9
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:43 ce
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:43 ee
-rw-r--r-- 1 xavi xavi 4.0K Sep 20 16:43 s3gw.db
-rw-r--r-- 1 xavi xavi  32K Sep 20 16:43 s3gw.db-shm
-rw-r--r-- 1 xavi xavi 1.4M Sep 20 16:43 s3gw.db-wal
```

WAL is 1.4M...

After stopping the `radosgw`:
```bash
$ ll /home/xavi/sfs
total 116K
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:43 34
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:43 3b
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:43 c9
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:43 ce
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:43 ee
-rw-r--r-- 1 xavi xavi 116K Sep 20 16:45 s3gw.db
```

**Comment by 0xavi0** [comment]
**More extra info**
This seems to happen ONLY for the multipart test.
I've renamed the multipart test:
```bash
$ mv test-sfs-multipart.py sfs-multipart.py
```
and after executing all the other integration tests:
```bash
$ python -m unittest test-*
```

The WAL file is also 4M

```bash
$ ll /home/xavi/sfs
total 4.4M
drwxr-xr-x 1 xavi xavi    8 Sep 20 16:55 00
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:56 08
drwxr-xr-x 1 xavi xavi    8 Sep 20 16:55 0a
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:56 0b
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 0c
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:56 0e
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 0f
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 12
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 13
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:56 14
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 15
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 16
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 1b
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:56 1c
drwxr-xr-x 1 xavi xavi    8 Sep 20 16:56 1d
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:56 1e
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:56 21
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:56 24
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:56 27
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 2a
drwxr-xr-x 1 xavi xavi    8 Sep 20 16:56 2b
drwxr-xr-x 1 xavi xavi    8 Sep 20 16:56 2e
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 2f
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:53 36
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 3b
drwxr-xr-x 1 xavi xavi    8 Sep 20 16:56 3c
drwxr-xr-x 1 xavi xavi    8 Sep 20 16:55 3d
drwxr-xr-x 1 xavi xavi   12 Sep 20 16:56 3f
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 41
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:56 43
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 46
drwxr-xr-x 1 xavi xavi    8 Sep 20 16:56 4e
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:56 52
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:56 54
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:56 56
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 58
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 5b
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:53 5e
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 60
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 61
drwxr-xr-x 1 xavi xavi   12 Sep 20 16:56 62
drwxr-xr-x 1 xavi xavi   12 Sep 20 16:56 67
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:56 68
drwxr-xr-x 1 xavi xavi    8 Sep 20 16:56 69
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:56 6b
drwxr-xr-x 1 xavi xavi   12 Sep 20 16:56 6c
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 71
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:56 73
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:56 75
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 7d
drwxr-xr-x 1 xavi xavi    8 Sep 20 16:56 7f
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 80
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 82
drwxr-xr-x 1 xavi xavi    8 Sep 20 16:55 87
drwxr-xr-x 1 xavi xavi    8 Sep 20 16:56 88
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:56 8a
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 8c
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 8d
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:56 91
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:56 97
drwxr-xr-x 1 xavi xavi    8 Sep 20 16:56 98
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 9b
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:56 9f
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 a2
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 a5
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 a6
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:53 a8
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:56 ab
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 ae
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 b0
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 b4
drwxr-xr-x 1 xavi xavi    8 Sep 20 16:56 bc
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:53 c4
drwxr-xr-x 1 xavi xavi    8 Sep 20 16:56 c6
drwxr-xr-x 1 xavi xavi    8 Sep 20 16:56 c9
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:56 ca
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:56 cb
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:56 cc
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:56 cd
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 ce
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 db
drwxr-xr-x 1 xavi xavi    8 Sep 20 16:56 dc
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:53 dd
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 e1
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:56 e4
drwxr-xr-x 1 xavi xavi    8 Sep 20 16:55 e9
drwxr-xr-x 1 xavi xavi   12 Sep 20 16:56 eb
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 ec
drwxr-xr-x 1 xavi xavi    8 Sep 20 16:56 ef
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 f7
drwxr-xr-x 1 xavi xavi    8 Sep 20 16:53 f8
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:56 fc
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:56 fd
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:53 fe
drwxr-xr-x 1 xavi xavi    4 Sep 20 16:55 ff
-rw-r--r-- 1 xavi xavi 356K Sep 20 16:56 s3gw.db
-rw-r--r-- 1 xavi xavi  32K Sep 20 16:56 s3gw.db-shm
-rw-r--r-- 1 xavi xavi 4.0M Sep 20 16:57 s3gw.db-wal
```

But after stopping and starting the `radosgw` the issue is not happening...

**Comment by tserong** [comment]
WTF?

**Comment by tserong** [comment]
(To clarify my last comment, it's perfectly fine for checkpointing to happen continually _while_ there's writes ongoing - for it to happen continually later when there's nothing else happening is rather bizarre)

**Comment by tserong** [comment]
OK, well the good news is, the problem isn't caused by our new checkpoint mechanism. I applied this patch to sqlite:

```diff
--- sqlite-src-3430100.orig/src/main.c
+++ sqlite-src-3430100/src/main.c
@@ -2374,6 +2374,7 @@ int sqlite3WalDefaultHook(
 ){
   if( nFrame>=SQLITE_PTR_TO_INT(pClientData) ){
     sqlite3BeginBenignMalloc();
+    fprintf(stderr, "sqlite3WalDefaultHook running passive checkpoint (%d frames)\n", nFrame);
     sqlite3_wal_checkpoint(db, zDb);
     sqlite3EndBenignMalloc();
   }
```

...then I ran radosgw with `--rgw-sfs-wal-checkpoint-use-sqlite-default --rgw-sfs-wal-size-limit=-1` to make it use the sqlite defaults, ran the mulitpart test, stopped and restarted radosgw, and I get the following continuously:

```
sqlite3WalDefaultHook running passive checkpoint (1000 frames)
sqlite3WalDefaultHook running passive checkpoint (1000 frames)
sqlite3WalDefaultHook running passive checkpoint (1000 frames)
sqlite3WalDefaultHook running passive checkpoint (1000 frames)
sqlite3WalDefaultHook running passive checkpoint (1000 frames)
[...]
```

**Comment by 0xavi0** [comment]
Oh, nice that the new code is not the cause.
Good catch!!!

**Comment by tserong** [comment]
I ran it under gdb with a breakpoint on `sqlite_wal_hook_callback`, backtraced it, and discovered it was stuck in the garbage collector, which runs on startup. So I added some more logging:

```diff
diff --git a/src/rgw/driver/sfs/sfs_gc.cc b/src/rgw/driver/sfs/sfs_gc.cc
index f21ce5fe358..8fd6383178e 100644
--- a/src/rgw/driver/sfs/sfs_gc.cc
+++ b/src/rgw/driver/sfs/sfs_gc.cc
@@ -44,6 +44,8 @@ SFSGC::~SFSGC() {
 }
 
 int SFSGC::process() {
+  lsfs_dout(this, 10) << "SFSGC::process() entry" << dendl;
+
   // This is the method that does the garbage collection.
   initial_process_time = ceph_clock_now();
   perfcounter->inc(l_rgw_sfs_gc_count);
@@ -58,6 +60,7 @@ int SFSGC::process() {
             sfs_gc_process_exit_state::delete_pending_objects_data
         )
     );
+    lsfs_dout(this, 10) << "SFSGC::process() delete_pending_objects_data returned false" << dendl;
     return 0;
   }
   // now delete possible pending multiparts data
@@ -69,6 +72,7 @@ int SFSGC::process() {
             sfs_gc_process_exit_state::delete_pending_multiparts_data
         )
     );
+    lsfs_dout(this, 10) << "SFSGC::process() delete_pending_multiparts_data returned false" << dendl;
     return 0;
   }
   // process deleted buckets
@@ -79,6 +83,7 @@ int SFSGC::process() {
         static_cast<uint64_t>(sfs_gc_process_exit_state::process_deleted_buckets
         )
     );
+    lsfs_dout(this, 10) << "SFSGC::process() process_deleted_buckets returned false" << dendl;
     return 0;
   }
   // process deleted objects
@@ -89,6 +94,7 @@ int SFSGC::process() {
         static_cast<uint64_t>(sfs_gc_process_exit_state::process_deleted_objects
         )
     );
+    lsfs_dout(this, 10) << "SFSGC::process() process_deleted_objects returned false" << dendl;
     return 0;
   }
 
@@ -98,6 +104,10 @@ int SFSGC::process() {
       l_rgw_sfs_gc_process_exit,
       static_cast<uint64_t>(sfs_gc_process_exit_state::finished)
   );
+  if (!time_to_process_more) {
+    lsfs_dout(this, 10) << "SFSGC::process() process_done_and_aborted_multiparts returned false" << dendl;
+  }
+  lsfs_dout(this, 10) << "SFSGC::process() exit" << dendl;
   return 0;
 }
 
@@ -147,6 +157,7 @@ bool SFSGC::process_deleted_buckets() {
                       << dendl;
   bool time_to_delete_more = true;
   for (auto const& bucket_id : deleted_buckets) {
+    lsfs_dout(this, 10) << "looping in SFSGC::process_deleted_buckets()" << dendl;
     time_to_delete_more = abort_bucket_multiparts(bucket_id);
     if (!time_to_delete_more) {
       return false;
@@ -275,6 +286,7 @@ bool SFSGC::delete_pending_multiparts_data() {
 }
 
 bool SFSGC::abort_bucket_multiparts(const std::string& bucket_id) {
+  lsfs_dout(this, 10) << "in SFSGC::abort_bucket_multiparts()" << dendl;
   common::PerfGuard elapsed(
       perfcounter, l_rgw_sfs_gc_abort_bucket_multiparts_elapsed
   );
@@ -293,6 +305,7 @@ bool SFSGC::abort_bucket_multiparts(const std::string& bucket_id) {
 bool SFSGC::delete_bucket_multiparts(
     const std::string& bucket_id, bool& all_parts_deleted
 ) {
+  lsfs_dout(this, 10) << "in SFSGC::delete_bucket_multiparts()" << dendl;
   sqlite::SQLiteMultipart db_mp(store->db_conn);
   pending_multiparts_to_delete = db_mp.remove_multiparts_by_bucket_id_transact(
       bucket_id, max_objects_to_delete_per_iteration
@@ -303,6 +316,7 @@ bool SFSGC::delete_bucket_multiparts(
 }
 
 bool SFSGC::delete_bucket(const std::string& bucket_id, bool& bucket_deleted) {
+  lsfs_dout(this, 10) << "bucket_id: " << bucket_id << ", deleted: " << bucket_deleted << dendl;
   sqlite::SQLiteBuckets db_buckets(store->db_conn);
   // deletes the db bucket (and all it's objects and versions) first in a
   // transaction.
diff --git a/src/rgw/driver/sfs/sqlite/sqlite_buckets.cc b/src/rgw/driver/sfs/sqlite/sqlite_buckets.cc
index 42877f50b28..a10f1f9ef58 100644
--- a/src/rgw/driver/sfs/sqlite/sqlite_buckets.cc
+++ b/src/rgw/driver/sfs/sqlite/sqlite_buckets.cc
@@ -167,6 +167,8 @@ std::optional<DBDeletedObjectItems> SQLiteBuckets::delete_bucket_transact(
             e.code().value() != SQLITE_CONSTRAINT) {
           throw(e);
         }
+        // TODO: how do we get real logging in here?
+        std::cerr << "SQLiteBuckets::delete_bucket_transact caught and ignored " << e.what() << std::endl;
       }
     }
     // try to delete the bucket
@@ -178,6 +180,8 @@ std::optional<DBDeletedObjectItems> SQLiteBuckets::delete_bucket_transact(
           e.code().value() != SQLITE_CONSTRAINT) {
         throw(e);
       }
+      // TODO: same question
+      std::cerr << "SQLiteBuckets::delete_bucket_transact caught and ignored " << e.what() << std::endl;
     }
     storage.commit();
     return ret_values;
```

Then I ran it again, and got this:

```
023-09-21T21:00:18.953+1000 7fd281c71f40  0 ceph version s3gw-v0.20.0-rc2-47-gdb90a42d514 (db90a42d514dfb0f1e78cd949cc1bd8dc01b8352) reef (dev), process radosgw, pid 313579
2023-09-21T21:00:18.953+1000 7fd281c71f40  0 framework: beast
2023-09-21T21:00:18.953+1000 7fd281c71f40  0 framework conf key: port, val: 7480
2023-09-21T21:00:18.953+1000 7fd281c71f40  1 init_numa not setting numa affinity
2023-09-21T21:00:18.953+1000 7fd281c71f40  0 rgw main: sfs init!
2023-09-21T21:00:18.953+1000 7fd27e4566c0 10 entry start
2023-09-21T21:00:18.956+1000 7fd281c71f40 -1 [SQLITE] (283) recovered 676 frames from WAL file /scratch/s3gw/qa/s3gw.db-wal
2023-09-21T21:00:18.956+1000 7fd281c71f40 10 db user version: 4
2023-09-21T21:00:18.993+1000 7fd281c71f40  0 sfs serving data from "/scratch/s3gw/qa"
2023-09-21T21:00:18.993+1000 7fd281c71f40  0 rgw main: Found default user "testid" in database.
2023-09-21T21:00:18.993+1000 7fd281c71f40 10 rgw main: initialize
2023-09-21T21:00:18.993+1000 7fd27cbf86c0  2 garbage collection: > GCWorker::entry start
2023-09-21T21:00:18.993+1000 7fd27cbf86c0 10 garbage collection: > SFSGC::process SFSGC::process() entry
2023-09-21T21:00:18.993+1000 7fd27abf46c0  5 lifecycle: schedule life cycle next start time: Thu Sep 21 14:00:00 2023
2023-09-21T21:00:18.993+1000 7fd278bf06c0  5 lifecycle: schedule life cycle next start time: Thu Sep 21 14:00:00 2023
2023-09-21T21:00:18.993+1000 7fd276bec6c0  5 lifecycle: schedule life cycle next start time: Thu Sep 21 14:00:00 2023
2023-09-21T21:00:18.993+1000 7fd281c71f40 10 get_sync_module: TODO
2023-09-21T21:00:18.993+1000 7fd27cbf86c0 10 garbage collection: > SFSGC::process_deleted_buckets deleted buckets found = 7
2023-09-21T21:00:18.993+1000 7fd27cbf86c0 10 garbage collection: > SFSGC::process_deleted_buckets looping in SFSGC::process_deleted_buckets()
2023-09-21T21:00:18.993+1000 7fd27cbf86c0 10 garbage collection: > SFSGC::abort_bucket_multiparts in SFSGC::abort_bucket_multiparts()
2023-09-21T21:00:18.993+1000 7fd281c71f40 10 get_lua_manager: TODO
2023-09-21T21:00:18.993+1000 7fd281c71f40  0 framework: beast
2023-09-21T21:00:18.993+1000 7fd281c71f40  0 framework conf key: ssl_certificate, val: config://rgw/cert/$realm/$zone.crt
2023-09-21T21:00:18.993+1000 7fd281c71f40  0 framework conf key: ssl_private_key, val: config://rgw/cert/$realm/$zone.key
2023-09-21T21:00:18.993+1000 7fd27cbf86c0 10 garbage collection: > SFSGC::delete_bucket_multiparts in SFSGC::delete_bucket_multiparts()
2023-09-21T21:00:18.993+1000 7fd27cbf86c0 10 garbage collection: > SFSGC::delete_bucket bucket_id: gsaauwvx.1695279451882282127, deleted: 0
SQLiteBuckets::delete_bucket_transact caught and ignored FOREIGN KEY constraint failed: constraint failed
2023-09-21T21:00:18.996+1000 7fd27cbf86c0 10 garbage collection: > SFSGC::delete_bucket bucket_id: gsaauwvx.1695279451882282127, deleted: 0
SQLiteBuckets::delete_bucket_transact caught and ignored FOREIGN KEY constraint failed: constraint failed

```

Those last lines (`SFSGC::delete_bucket bucket_id: gsaauwvx.1695279451882282127, deleted: 0` and `QLiteBuckets::delete_bucket_transact caught and ignored FOREIGN KEY constraint failed: constraint failed`) repeat forever, punctuated by the occasional `sqlite3WalDefaultHook running passive checkpoint (1000 frames)`.

This raises two questions:

- Why are we ignoring/suppressing constraint errors when attempting to delete buckets?
- What is it about running the multipart test that causes s3gw to get into a state where bucket deletes cause constraint errors?

**Comment by tserong** [comment]
I've just tweaked my crappy logging slightly:

```diff
diff --git a/src/rgw/driver/sfs/sqlite/sqlite_buckets.cc b/src/rgw/driver/sfs/sqlite/sqlite_buckets.cc
index 42877f50b28..e615af39d8e 100644
--- a/src/rgw/driver/sfs/sqlite/sqlite_buckets.cc
+++ b/src/rgw/driver/sfs/sqlite/sqlite_buckets.cc
@@ -167,6 +167,9 @@ std::optional<DBDeletedObjectItems> SQLiteBuckets::delete_bucket_transact(
             e.code().value() != SQLITE_CONSTRAINT) {
           throw(e);
         }
+        // TODO: how do we get real logging in here?
+        std::cerr << "Trying to remove object " << uuid_version << std::endl;
+        std::cerr << "SQLiteBuckets::delete_bucket_transact caught and ignored " << e.what() << std::endl;
       }
     }
     // try to delete the bucket
@@ -178,6 +181,9 @@ std::optional<DBDeletedObjectItems> SQLiteBuckets::delete_bucket_transact(
           e.code().value() != SQLITE_CONSTRAINT) {
         throw(e);
       }
+      // TODO: same question
+      std::cerr << "Trying to remove bucket " << bucket_id << std::endl;
+      std::cerr << "SQLiteBuckets::delete_bucket_transact caught and ignored " << e.what() << std::endl;
     }
     storage.commit();
     return ret_values;
```

This confirms it's attempted bucket deletion, not object deletion, that causes the foreign key error:

```
2023-09-21T21:54:53.458+1000 7f68a55716c0 10 garbage collection: > SFSGC::delete_bucket bucket_id: pnhxguof.1695296867792266931, deleted: 0
Trying to remove bucket pnhxguof.1695296867792266931
SQLiteBuckets::delete_bucket_transact caught and ignored FOREIGN KEY constraint failed: constraint failed
```

**Comment by tserong** [comment]
Messing with the sqlite cli after running the multipart integration test:

```
sqlite> delete from buckets;
sqlite> PRAGMA foreign_key_check;
multiparts|1|buckets|0
multiparts|2|buckets|0
multiparts|3|buckets|0
multiparts|4|buckets|0
```

I'll try moving some things around inside `SFSGC::process()`, e.g. shift the call to `process_deleted_buckets()` to the very end, after the call to `process_done_and_aborted_multiparts()`

**Comment by 0xavi0** [comment]
I think that's, indeed, the problem
The multipart integration tests deletes the bucket. 
Then on the next round GC triggers and tries to delete the bucket before removing the done and aborted multiparts.
If finds done and aborted multiparts in a deleted bucket, but code is trying to delete the bucket first, which is breaking the foreign keys constraint .
I think moving `process_done_and_aborted_multiparts()` to be done BEFORE deleting buckets should fix the problem here but the main problem is that `SFSGC::delete_bucket` is calling a transaction that can be retried and it's not checking that the return value is now std::nullopt (which would mean the transaction was not correct).

As `SFSGC::delete_bucket`  is called in a loop that is never deleting the bucket and that's why `GC` gets stuck in there.

Also, I think CONSTRAINT errors should not be ignored in the `RetrySQLite` class.
At this code:
```c++
std::optional<Return> run() {
    if (perfcounter) {
      perfcounter->inc(l_rgw_sfs_sqlite_retry_total, 1);
    }
    for (int retry = 0; retry < m_max_retries; retry++) {
      try {
        Return result = m_fn();
        m_successful = true;
        m_failed_error = std::error_code{};
        m_retries = retry;
        return result;
      } catch (const std::system_error& ex) {
        m_failed_error = ex.code();
        if (critical_error(ex.code().value())) {
          // Rethrow, expect a higher layer to shut us down
          throw ex;
        }
        std::this_thread::sleep_for(10ms * retry);
        m_retries = retry;
        if (perfcounter) {
          perfcounter->inc(l_rgw_sfs_sqlite_retry_retried_count, 1);
        }
      }
    }
    m_successful = false;
    if (perfcounter) {
      perfcounter->inc(l_rgw_sfs_sqlite_retry_failed_count, 1);
    }
    return std::nullopt;
  };
```
it calls `critical_error(ex.code().value())`, but `SQLITE_CONSTRAINT` is not in the list:
```c++
bool critical_error(int ec) {
  switch (ec) {
    case SQLITE_ERROR:
    case SQLITE_INTERNAL:
    case SQLITE_PERM:
    case SQLITE_NOMEM:
    case SQLITE_READONLY:
    case SQLITE_IOERR:
    case SQLITE_CORRUPT:
    case SQLITE_NOTFOUND:
    case SQLITE_FULL:
    case SQLITE_CANTOPEN:
    case SQLITE_PROTOCOL:
    case SQLITE_TOOBIG:
    case SQLITE_MISMATCH:
    case SQLITE_MISUSE:
    case SQLITE_NOLFS:
    case SQLITE_AUTH:
    case SQLITE_RANGE:
    case SQLITE_NOTADB:
      return true;
    default:
      return false;
  }
}
```

**Comment by 0xavi0** [comment]
After debugging this I found out this was caused by multiparts with 0 parts. 

Code was not taking those into account and was considering the bucket eligible to be deleted. 

When trying to delete the bucket, it throws a `SQLITE_CONSTRAINT` error, which is ignored because the `SQLiteBuckets::delete_bucket_transact` transaction itself could potentially throw the exception as it is implemented following the _Easier to ask for forgiveness than permission_ approach.  (As discussed with the team we concluded that was the approach to follow).

Problem here is that we were shooting at out own feet due to this as the constraint was broken not because of the objects and versions inside a bucket, but by the multiparts related to it. 

A different possible solution could be to mix multiparts-parts with objects-versions deletion in the `SQLiteBuckets::delete_bucket_transact`, but I think that makes the whole implementation more complex because we would need to return different entities from the `SQLiteBuckets::delete_bucket_transact` call. 
Maybe we could mix them in a future iteration returning a base class with a common interface that simplifies things, but I thought that was too complex now to fix this asap. 

https://github.com/aquarist-labs/ceph/pull/218 tries to fix the issue by taking the multiparts with 0 parts into account.
